R/repr_matrix_df.r

```diff
@@ -47,78 +47,79 @@ ellip.limit.vec <- function(v, num, ellip) {
 	c(v[lims$begin], ellip, v[lims$end])
 }
 
+# returns a character array with optionally a section of columns and rows in the middle replaced by ellipses
 ellip.limit.arr <- function(
 	a,
 	rows = getOption('repr.matrix.max.rows'),
 	cols = getOption('repr.matrix.max.cols')
 ) {
 	stopifnot(rows >= 2L, cols >= 2L)
-
-	# Don't worry about any of the code below if the array is already small.
-	if (rows >= nrow(a) && cols >= ncol(a)) {
-		return(a)
-	}
-
-	left   <- seq_len(ceiling(cols / 2))
-	right  <- seq.int(ncol(a) - floor(cols / 2) + 1L, ncol(a))
-	top    <- seq_len(ceiling(rows / 2))
-	bottom <- seq.int(nrow(a) - floor(rows / 2) + 1L, nrow(a))
 	
-	# fix columns that won't like ellipsis being inserted
-	if (is.data.frame(a)) {
-		# data.tables can't be indexed by column number, unless you provide the
-		# with=FALSE parameter. To avoid the hassle, just convert to a normal table.
-		# dplyr's tbl_* objects don't collapse to a vector when indexed by column,
-		# so functions like 'is.factor' always return false. Again, just drop to a
-		# basic data.table and avoid the hassle.
-		if (inherits(a, c('data.table', 'tbl'))) {
-			a <- as.data.frame(a)
-		}
-		for (c in seq_len(ncol(a))) {
-			if (is.factor(a[, c])) {
-				# Factors: add ellipses to levels
-				levels(a[, c]) <- c(levels(a[, c]), ellipses)
-			} else if (inherits(a[, c], 'Date') || inherits(a[, c], 'POSIXt') ) {
-				# Dates: convert to plain strings
-				a[, c] <- as.character(a[, c])
-			}
-		}
+	many_rows <- rows < nrow(a)
+	many_cols <- cols < ncol(a)
+	
+	# create sequences of indices to bisect rows and columns
+	if (many_rows) {
+		upper <- seq_len(ceiling(rows / 2))
+		lower <- seq.int(nrow(a) - floor(rows / 2) + 1L, nrow(a))
 	}
-
-	if (rows < nrow(a) && inherits(a, 'tbl')) {
-		# tbl objects from dplyr automatically reset their row names when sliced.
-		# we'd like to make it clear that the array has had rows cut out, so that
-		# behavior isn't ideal here. If we're row-slicing a tbl object, first 
-		# convert to an ordinary data.frame.
-		a <- as.data.frame(a)
+	if (many_cols) {
+		left  <- seq_len(ceiling(cols / 2))
+		right <- seq.int(ncol(a) - floor(cols / 2) + 1L, ncol(a))
 	}
 	
-	if (rows < nrow(a) && cols < ncol(a)) {
-		if (is.matrix(a)) {
-			# If a is a matrix, R will coerce the factor to character and sub in 
-			# the factor *value*, not the level.  You end up with a bunch of 1s.
-			ehf <- ellip.h
+	# assign a list of parts that can be coerced to strings
+	limitation <- if (many_rows && many_cols) {
+		parts <- list(
+			ul = a[upper, left], ur = a[upper, right],
+			ll = a[lower, left], lr = a[lower, right])
+		'both'
+	} else if (many_rows) {
+		parts <- list(
+			upper = a[upper, , drop = FALSE],
+			lower = a[lower, , drop = FALSE])
+		'vert'
```

Comment:
can we keep the "many_rows" naming? I know `vert` will align the source code, but on the other hand I can see two interpretations for "vertical": "slice it vertically" or "too many stuff in the vertical dimension"... `too_many_rows` is without ambiguity.